File: tools/typescript/src/ai-sdk/toolkit.ts
import {tool, Tool} from 'ai';
import {z} from 'zod';
import {jsonSchemaToZod} from '../shared/schema-utils';
import {ToolkitCore, ToolkitConfig} from '../shared/toolkit-core';
import type {
  LanguageModelV2Middleware,
  LanguageModelV2Usage,
} from '@ai-sdk/provider';

type ProviderTool = Tool<any, any>;
type WrapGenerateOptions = Parameters<
  NonNullable<LanguageModelV2Middleware['wrapGenerate']>
>[0];
type WrapStreamOptions = Parameters<
  NonNullable<LanguageModelV2Middleware['wrapStream']>
>[0];

type StripeMiddlewareConfig = {
  billing?: {
    type?: 'token';
    customer: string;
    meters: {
      input?: string;
      output?: string;
    };
  };
};

class StripeAgentToolkit {
  private _core: ToolkitCore;
  private _tools: Record<string, ProviderTool> = {};

  /**
   * The tools available in the toolkit.
   * @deprecated Access tools via getTools() after calling initialize().
   * Direct property access will return empty object if not initialized.
   */
  get tools(): Record<string, ProviderTool> {
    this._core.warnIfNotInitialized();
    return this._tools;
  }

  constructor(config: ToolkitConfig) {
    this._core = new ToolkitCore(config);
  }

  /**
   * Initialize the toolkit by connecting to the MCP server.
   * Must be called before using tools.
   */
  initialize(): Promise<void> {
    return this._core.initialize((filteredTools) => {
      // Convert MCP tools to AI SDK format
      for (const remoteTool of filteredTools) {
        const zodSchema = jsonSchemaToZod(remoteTool.inputSchema);

        this._tools[remoteTool.name] = tool({
          description: remoteTool.description || remoteTool.name,
          inputSchema: zodSchema,
          execute: (args: z.infer<typeof zodSchema>) => {
            return this._core.stripe.run(remoteTool.name, args);
          },
        });
      }
    });
  }

  /**
   * Check if the toolkit has been initialized.
   */
  isInitialized(): boolean {
    return this._core.isInitialized();
  }

  /**
   * Middleware for billing based on token usage.
   * Note: This uses direct Stripe SDK calls, not MCP.
   */
  middleware(config: StripeMiddlewareConfig): LanguageModelV2Middleware {
    const bill = async (usage?: LanguageModelV2Usage) => {
      if (!config.billing || !usage) {
        return;
      }

      const {inputTokens, outputTokens} = usage;
      const inputValue = (inputTokens ?? 0).toString();
      const outputValue = (outputTokens ?? 0).toString();

      if (config.billing.meters.input) {
        await this._core.stripe.createMeterEvent({
          event: config.billing.meters.input,
          customer: config.billing.customer,
          value: inputValue,
        });
      }

      if (config.billing.meters.output) {
        await this._core.stripe.createMeterEvent({
          event: config.billing.meters.output,
          customer: config.billing.customer,
          value: outputValue,
        });
      }
    };

    return {
      wrapGenerate: async ({doGenerate}: WrapGenerateOptions) => {
        const result = await doGenerate();

        await bill(result.usage);

        return result;
      },

      wrapStream: async ({doStream}: WrapStreamOptions) => {
        const {stream, ...rest} = await doStream();

        const transformStream = new TransformStream<any, any>({
          async transform(chunk, controller) {
            if (chunk?.type === 'finish') {
              await bill(chunk.usage);
            }

            controller.enqueue(chunk);
          },
        });

        return {
          stream: stream.pipeThrough(transformStream),
          ...rest,
        };
      },
    };
  }

  /**
   * Get the tools. Throws if not initialized.
   */
  getTools(): Record<string, ProviderTool> {
    this._core.ensureInitialized();
    return this._tools;
  }

  /**
   * Close the toolkit connection and clean up resources.
   * Safe to call multiple times (idempotent).
   */
  close(): Promise<void> {
    return this._core.close(() => {
      this._tools = {};
    });
  }
}

/**
 * Factory function to create and initialize a StripeAgentToolkit.
 * Provides a simpler async initialization pattern.
 *
 * @example
 * const toolkit = await createStripeAgentToolkit({
 *   secretKey: 'rk_test_...',
 *   configuration: { actions: { customers: { create: true } } }
 * });
 * const tools = toolkit.getTools();
 */
export async function createStripeAgentToolkit(
  config: ToolkitConfig
): Promise<StripeAgentToolkit> {
  const toolkit = new StripeAgentToolkit(config);
  await toolkit.initialize();
  return toolkit;
}

export default StripeAgentToolkit;

File: tools/python/stripe_agent_toolkit/openai/toolkit.py
"""Stripe Agent Toolkit for OpenAI Agents SDK."""

import json
from typing import List, Optional, Dict, Any

from agents import FunctionTool
from agents.run_context import RunContextWrapper

from ..shared.toolkit_core import ToolkitCore
from ..shared.mcp_client import McpTool
from ..configuration import Configuration
from .hooks import BillingHooks


class StripeAgentToolkit(ToolkitCore[List[FunctionTool]]):
    """
    Stripe Agent Toolkit for OpenAI Agents SDK.

    Example:
        toolkit = await create_stripe_agent_toolkit(
            secret_key='rk_test_...',
            configuration={'actions': {'customers': {'create': True}}}
        )
        tools = toolkit.get_tools()
        agent = Agent(name="Stripe Agent", tools=tools)
        await toolkit.close()
    """

    def __init__(
        self,
        secret_key: str,
        configuration: Optional[Configuration] = None
    ):
        super().__init__(secret_key, configuration)

    def _empty_tools(self) -> List[FunctionTool]:
        """Return empty list of tools."""
        return []

    def _convert_tools(
        self,
        mcp_tools: List[McpTool]
    ) -> List[FunctionTool]:
        """Convert MCP tools to OpenAI FunctionTool instances."""
        tools = []
        for mcp_tool in mcp_tools:
            tools.append(self._create_function_tool(mcp_tool))
        return tools

    def _create_function_tool(self, mcp_tool: McpTool) -> FunctionTool:
        """Create a FunctionTool from an MCP tool definition."""
        stripe_client = self._stripe
        tool_name = mcp_tool["name"]

        async def on_invoke_tool(
            ctx: RunContextWrapper[Any],
            input_str: str
        ) -> str:
            args = json.loads(input_str)
            return await stripe_client.run(tool_name, args)

        # Prepare parameters schema
        parameters = dict(mcp_tool.get("inputSchema", {}))
        parameters["additionalProperties"] = False
        parameters["type"] = "object"

        # Clean up schema - remove fields not needed in OpenAI function schema
        for key in ["description", "title"]:
            parameters.pop(key, None)

        if "properties" in parameters:
            for prop in parameters["properties"].values():
                for key in ["title", "default"]:
                    if isinstance(prop, dict):
                        prop.pop(key, None)

        return FunctionTool(
            name=tool_name,
            description=mcp_tool.get("description", tool_name),
            params_json_schema=parameters,
            on_invoke_tool=on_invoke_tool,
            strict_json_schema=False
        )

    @property
    def tools(self) -> List[FunctionTool]:
        """
        The tools available in the toolkit.

        .. deprecated::
            Access tools via get_tools() after calling initialize().
        """
        return self._get_tools_with_warning()

    def billing_hook(
        self,
        type: Optional[str] = None,
        customer: Optional[str] = None,
        meter: Optional[str] = None,
        meters: Optional[Dict[str, str]] = None
    ) -> BillingHooks:
        """
        Create a billing hook for usage metering.

        Args:
            type: Type of billing - "outcome" or "token"
            customer: Stripe customer ID
            meter: Single meter event name for outcome-based billing
            meters: Dict with 'input' and 'output' meter names for token billing

        Returns:
            BillingHooks instance for use with agents
        """
        return BillingHooks(self._stripe, type, customer, meter, meters)


async def create_stripe_agent_toolkit(
    secret_key: str,
    configuration: Optional[Configuration] = None
) -> StripeAgentToolkit:
    """
    Factory function to create and initialize a StripeAgentToolkit.

    This is the recommended way to create a toolkit as it handles
    async initialization automatically.

    Example:
        toolkit = await create_stripe_agent_toolkit(
            secret_key='rk_test_...',
            configuration={'actions': {'customers': {'create': True}}}
        )
        tools = toolkit.get_tools()

        # Use with agent
        agent = Agent(name="Stripe Agent", tools=tools)

        # Clean up when done
        await toolkit.close()

    Args:
        secret_key: Stripe API key (rk_* recommended, sk_* deprecated)
        configuration: Optional configuration for actions and context

    Returns:
        Initialized StripeAgentToolkit ready to use
    """
    toolkit = StripeAgentToolkit(secret_key, configuration)
    await toolkit.initialize()
    return toolkit
